Update to Hector 3

rgieseke opened the pull request:
- [x] fix TODOs
- [x] update and check tests/comparisons with vanilla Hector
- [x] finish conversion of variable renamings, see http://jgcri.github.io/hector/articles/v3_output_changes.html
- [x] fix versioneer config

rgieseke committed:
WIP Update to Hector 3

rgieseke commented:
Test script

```py
import pyhector
from copy import deepcopy
from pyhector.default_config import _default_config

print(pyhector.__hector_version__)

h = pyhector.Hector()

h.set_emissions(pyhector.ssp126)

parameters = deepcopy(_default_config)
h.config(parameters)

h.run(until=2100)
```

Prints the expected Hector version and runs.

However, adding an observable fails:

```py
import pyhector
from copy import deepcopy
from pyhector.default_config import _default_config

print(pyhector.__hector_version__)

h = pyhector.Hector()

h.set_emissions(pyhector.ssp126)

parameters = deepcopy(_default_config)
h.config(parameters)

h.add_observable(
    "simpleNbox",
    "atmos_co2",
    False
)

h.run(until=2100)
```

Error message:
```py
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/Work/pyhector/session.py:20
     12 h.config(parameters)
     14 h.add_observable(
     15     "simpleNbox",
     16     "atmos_co2",
     17     False
     18 )
---> 20 h.run(until=2100)

ValueError: vector::_M_fill_insert
```

rgieseke committed:
Update workflow Python version

rgieseke committed:
Add pyproject.toml and update versioneer

Versioneer handling needs to be sorted out. Duplicated in setup.cfg.

swillner committed:
Add missing versionfile_source to pyproject.toml

swillner committed:
Use hcore->getCurrentDate instead of date from shouldVisit as the latter changed to relative date

swillner commented:
(hint for enabling debug compilation: add to `Pybind11Extension` in `setup.py`: `extra_compile_args=["-g3"]`)

rgieseke committed:
Update to Hector 3.2.0

rgieseke commented:
With the now compiling and running version only a single value is returned in the array.
 
```python
import pyhector
from copy import deepcopy
from pyhector.default_config import _default_config

print(pyhector.__hector_version__)

h = pyhector.Hector()

h.set_emissions(pyhector.ssp126)

parameters = deepcopy(_default_config)
h.config(parameters)

h.add_observable(
	"simpleNbox",
	"CO2_concentration",
	True
)

h.run()

out = h.get_observable("simpleNbox", "CO2_concentration")

print(out)
```

swillner committed:
Fix time handling

rgieseke committed:
Update index after wrapper fixes to include startDate

rgieseke committed:
Add needs_date where necessary

rgieseke committed:
Fix warnings

rgieseke committed:
Adjust read_hector_input for SSPs

rgieseke committed:
Add SSPs and update test_write_hector_input

rgieseke committed:
Update SSP tests

rgieseke committed:
Update required dates

rgieseke committed:
Remove deprecated RCP test data

rgieseke committed:
Fix tests

Still has some Todos.

rgieseke committed:
Add SSP outputstream files

rgieseke committed:
Fix test

rgieseke committed:
Blacken

rgieseke committed:
Update ocean and temperature outputs

rgieseke committed:
Remove RCP scenarios

rgieseke committed:
Update defaults

rgieseke committed:
Update plot

rgieseke committed:
Fix key duplication

rgieseke committed:
Fix import

rgieseke committed:
Lint and blacken

rgieseke committed:
Update to latest Black version

rgieseke committed:
Fix import order

rgieseke committed:
Update Python and Action versions in CI

rgieseke committed:
Fix import order

rgieseke committed:
Edit Readme

rgieseke committed:
Update years and email

rgieseke committed:
Update years and doc links

rgieseke committed:
Doc and code tweaks

rgieseke committed:
Update changelog

rgieseke committed:
Adjust doc building check

rgieseke committed:
Only run one sphinx build command

rgieseke committed:
Revert "Only run one sphinx build command"

This reverts commit 0451aeae9797f5c101d829af00d7c3f41b763566.

rgieseke committed:
Remove documentation coverage check

rgieseke committed:
Test series similarity by comparing significant digits

rgieseke committed:
Add albedo radiative forcing from scenario input file

rgieseke committed:
Remove albedo from comparison as not part of base config

rgieseke committed:
Update config

rgieseke committed:
Update plot

rgieseke committed:
Remove old ini file

rgieseke committed:
Update comment

rgieseke commented:
Pyhector v2.x results and output streams from the C++ command line version are in good agreement, but the tests are not catching larger differences, like volcanic forcing changes, missing albedo forcing, or parameter changes introduced with the Hector v3 update.

```python
import pyhector

df = pyhector.run(pyhector.rcp26)

outstream = pyhector.read_hector_output(
    "https://github.com/openclimatedata/pyhector/raw/main/tests/data/outputstream_rcp26.csv"
)

df["temperature.Tgav"].plot(color="orange")
outstream["Tgav"].plot(linestyle=":", color="blue")
```

![image](https://github.com/openclimatedata/pyhector/assets/198537/0c2de61e-1735-4ff2-bf7f-ff3e582f8041)

I think the Pandas testing function uses numpy's `isclose` which actually does
```python
absolute(a - b) <= (atol + rtol * absolute(b))
```

The test only changes the `rtol` default.
```python
assert_series_equal(
    output["temperature.global_tas"],
    original.global_tas,
    check_names=False,
    check_exact=False,
    rtol=3,
)
 ```

rgieseke commented:
I'm not entirely sure how command line Hector sets its output precision for the CSV file, but it appears to have four significant figures.
One way to make a comparison between Pyhector output and the CSV streams could be to compare the difference between these significant figures, ignoring the exponent and checking that there are only floating point rounding differences.
```python
import os
import pyhector
from pyhector import (
    ssp119,
    ssp126,
    ssp245,
    ssp370,
    ssp434,
    ssp460,
    ssp534_over,
    ssp585,
    read_hector_output
)

import matplotlib.pyplot as plt


ssps = {
    "ssp119": ssp119,
    "ssp126": ssp126,
    "ssp245": ssp245,
    "ssp370": ssp370,
    "ssp434": ssp434,
    "ssp460": ssp460,
    "ssp534-over": ssp534_over,
    "ssp585": ssp585,
}


for name, scenario in list(ssps.items())[-2:-1]:
    df = pyhector.run(scenario)
    out = read_hector_output(os.path.join("tests/data/outputstream_{}.csv".format(name)))
    
    fig, ax = plt.subplots(1,3, figsize=(14,6))
    out["global_tas"].plot(label="cli outputstream", ax=ax[0])
    df["temperature.global_tas"].plot(label="pyhector", ax=ax[0])
    ax[0].legend()
    ax[0].set_title("Temperature")
    
    out["CO2_concentration"].plot(label="cli outputstream", ax=ax[1])
    df["simpleNbox.CO2_concentration"].plot(label="pyhector", ax=ax[1])
    ax[1].legend()
    ax[1].set_title("Concentration")

    (out["global_tas"].map(lambda x: float(f'{x:.5e}'[:7])) - 
     df["temperature.global_tas"].map(lambda x: float(f'{x:.5e}'[:7]))).plot(ax=ax[2])
    ax[2].set_title('Subtracting significant digits (Temp.)')
    
    fig.suptitle(name)

    assert (out["global_tas"].map(lambda x: float(f'{x:.5e}'[:7])) - 
        df["temperature.global_tas"].map(lambda x: float(f'{x:.5e}'[:7])).abs() < 6e-4).all()
```

![image](https://github.com/openclimatedata/pyhector/assets/198537/bb3d9e75-4a57-41ab-bea5-7fb70fef7e95)

This should catch small changes, e.g. like 
```python
df = pyhector.run(scenario, {"temperature": {"S": 3.01}})
```

rgieseke committed:
Enable simpleNbox.NBP output

See also https://github.com/JGCRI/hector/issues/735

rgieseke committed:
Update outputs

Variables in outputstream should be covered.
A few (revelle and slr) don't work yet.

rgieseke committed:
Add test for checking output units

rgieseke committed:
Order dictionary by key

rgieseke committed:
Adjust unit to one used in outputstream

rgieseke committed:
Update versioneer

swillner commented on include/Hector.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2018 Sven Willner <sven.willner@pik-potsdam.de>
+ * Copyright (c) 2017 - 2024 Sven Willner <sven.willner@gmail.com>
```
```suggestion
 * Copyright (c) 2017-2024 Sven Willner <sven.willner@gmail.com>
```

swillner commented on include/Observable.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Sven Willner <sven.willner@pik-potsdam.de>
+ * Copyright (c) 2018 - 2024 Sven Willner <sven.willner@gmail.com>
```
```suggestion
 * Copyright (c) 2018-2024 Sven Willner <sven.willner@gmail.com>
```

swillner commented on src/main.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Sven Willner <sven.willner@pik-potsdam.de>
+ * Copyright (c) 2018 - 2024 Sven Willner <sven.willner@gmail.com>
```
```suggestion
 * Copyright (c) 2018-2024 Sven Willner <sven.willner@gmail.com>
```

swillner reviewed (approved):
awesome!

rgieseke committed:
Formatting

Co-authored-by: Sven Willner <sven.willner@gmail.com>

rgieseke committed:
Formatting

Co-authored-by: Sven Willner <sven.willner@gmail.com>

rgieseke committed:
Formatting

Co-authored-by: Sven Willner <sven.willner@gmail.com>
